src/Search/Services/IndexableService.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace SilverStripe\FullTextSearch\Search\Services;
+
+use SilverStripe\Core\Extensible;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * Checks if a DataObject is publically viewable thus able to be added or retrieved from a publically searchable index
+ * Caching results because these checks may be done multiple times as there a few different code paths that search
+ * results might follow in real-world search implementations
+ */
+class IndexableService
+{
+    use Injectable;
+    use Extensible;
+
+    protected $cache = [];
+
+    public function clearCache(): void
+    {
+        $this->cache = [];
+    }
+
+    public function isIndexable(DataObject $obj): bool
```

Comment:
So this wouldn't work if e.g. one language variation of an object is indexable but the other one isn't. Funnily enough this PR from Terraformers might come in handy for this: https://github.com/silverstripe/silverstripe-framework/pull/9400. For now, it's good enough

Reply:
you could modify this via `$this->extend('updateIsIndexable', $obj, $value);` ?

e.g.
```
public function updateIsIndexable(&$obj, &$value) {
   if ($obj->IsEnglish) {
    $value = true;
   } elseif ($obj->IsFrench) {
    $value = false;
  } 
}
```
?

Reply:
Though, I guess if you have $obj->EnglishContent and $obj->FrenchContent, then it won't be so good

Would it be better if we cache by default and have a Configurable opt-out?  Would mean that you'd need to do (apologies if I have syntax wrong) $this->config()->get('do_cache_is_indexable') on every check, which would lead to some level of performance degredation

Reply:
Is this worth addressing upfront, rather than waiting for it to impact a project? Do we have a gauge on how Fluent is currently used with fulltextsearch – do projects use separate indexes for each locale?

Reply:
There's not even a search variant class for Fluent (like for [subsites](https://github.com/silverstripe/silverstripe-fulltextsearch/blob/3/src/Search/Variants/SearchVariantSubsites.php)).

I assume that when you call reindex on a site with fluent, you already have to perform some customisations to that indexing logic.  We could just reset the cache somewhere around here? https://github.com/silverstripe/silverstripe-fulltextsearch/blob/277a307a1b225587e93f69d86c3f5f2817a97f3d/src/Solr/Reindex/Handlers/SolrReindexBase.php#L55

Reply:
This might also be an argument *against* employing a cache here. Can't we just ensure that isIndexable() is only called once per record/variant on an indexing request? ShowInSearch is cheap, canView is expensive. So worth caching, but only if it's called multiple times.

Reply:
>>  Can't we just ensure that isIndexable() is only called once per record/variant on an indexing request

- Outlet filtering - we cannot ensure isIndexable() is only called once
- Inlet filtering we can be fairly confident that it's only being called once

Outlet filtering is messy because that's where we see the most variation with custom implementations.  Because of this we've needed to put isIndexable() is multiple places because there is central place (that's still sane) that all requests in all custom implementations from the solr server go through

e.g.
- Scenario 1. Uses SearchForm.php, CustomSolrIndex->search() calls parent::search() - isIndexable() called twice
- Scenario 2. Does not use SearchForm.php, instead calls CustomSolrIndex->search() from in a page template (probably did not follow docs :-)) - isIndexable() called once
- Scenario 3. Uses SearchForm.php, CustomSolrIndex->search() is copy pasted from SolrIndex, does not call parent::search() - isIndexable() called once

Inlet filtering on the other hand is much easier since it's either SearchUpdateProcessor (cms->save() or $dataObject->write() or SolrReindexBase (dev/tasks/Solr_Reindex), bother of which are very unlikely to be customised

A good argument to remove caching from the service would be that it's only there for the sake of outlet filtering, there it's likely to be paginated to say 10 results, so you're looking at 20 vs 10 canView() calls, so the performance hit may not be too bad

I can see a few options here:
a) Remove caching from the service and accept the outlet filtering performance hit (most secure and easiest)
b) Add the Configurable trait and make caching opt-out so that modules such as fluent can disable it (though those modules will themselves need to be tested and updated, creates a lot of work)
c) Test for the existence modules such as fluent from a hardcoded list within this module and if class_exists(Fluent), then we disable caching
d) Keep things as they are now, accept some level of risk that isIndexable() will return the wrong result in some scenarios we haven't yet though of.

I don't really see removing isIndexable() from multiple locations as a valid option, just because custom search implementations didn't quite do things by the book doesn't seem like a good enough justification to shift responsibility to them for things going wrong.

I don't like b) or c).  Too complicated.

a) is attractive because it'll just work, will result in front end search results being slower though
d) is probably fine.  Inlet filtering will largely prevent anything that shouldn't be there getting indexed in the first place.  Outlet filtering is the second layer of defense.

Let me know what option you think is best.

Reply:
Hey Steve, thanks for working through this, great TED talk would read again ;)
How about a separate `getCacheKey($obj)` which can then be augmented by `updateCacheKey($obj)`? The getter then simply concats any additional cache keys. In some cases cache keys in extensions will be dependant on wider context (versioned stage, active fluent locale), but any extension can read this context. We wouldn't have to implement it for Fluent etc, just ensure that it *can* be done. This is somewhat of a temporary API anyway and should be superseded with `DataObject->getUniqueKey()` once that's merged (https://github.com/silverstripe/silverstripe-framework/pull/9400). But we need to ensure backwards compatibility here.

Reply:
Sounds good.  Have implemented.  I'll mark this one as resolved.